tools/federation/src/main/java/org/eclipse/rdf4j/federated/structures/FedXTupleQuery.java

```diff
@@ -48,8 +51,26 @@ public TupleQueryResult evaluate() throws QueryEvaluationException {
 	@Override
 	public void evaluate(TupleQueryResultHandler handler)
 			throws QueryEvaluationException, TupleQueryResultHandlerException {
+
+		// attach the handler to the query to allow pass through of results
+		// for single source queries
+		TupleExpr tupleExpr = getParsedQuery().getTupleExpr();
+		PassThroughTupleExpr passThroughTupleExpr = new PassThroughTupleExpr(tupleExpr,
+				handler);
+		delegate.getParsedQuery().setTupleExpr(passThroughTupleExpr);
+
 		FedXUtil.applyQueryBindings(this);
-		delegate.evaluate(handler);
+		TupleQueryResult tqr = delegate.evaluate();
+
+		if (!passThroughTupleExpr.isPassedThrough()) {
+			// if the result is not passed through to the handler directly,
+			// we need to make sure to report the result. Note that only
+			// SingleSourceQuery instances can be passed through
+			QueryResults.report(tqr, handler);
+		} else {
+			// to be absolutely sure that everything is closed
+			tqr.close();
+		}
```

Comment:
Should this be in a try-with-resources block rather than this if-else?

Reply:
I think this is not absolutely require, but it doesn't harm either.

In SailTupleQuery it is also

```
TupleQueryResult queryResult = evaluate();
QueryResults.report(queryResult, handler);
```

The only minimal difference here is the call to isPssedThrough which cannot throw an exception.

Do you still think we should have the close in a try with resources block?

Reply:
Nah, fair enough.